src/Mapping/Annotation/ForwardCompatibilityTrait.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Doctrine Behavioral Extensions package.
+ * (c) Gediminas Morkevicius <gediminas.morkevicius@gmail.com> http://www.gediminasm.org
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Gedmo\Mapping\Annotation;
+
+/**
+ * @todo Remove this trait when support for array based attributes is removed.
+ *
+ * @internal
+ */
+trait ForwardCompatibilityTrait
+{
+    /**
+     * If the passed value has a different value than the default one, it will be returned. Otherwise, it returns the attribute
+     * value from the `$data` array or `null` if it is not defined.
+     *
+     * @param mixed                $value
+     * @param array<string, mixed> $data
+     *
+     * @return mixed
+     */
+    private function getAttributeValue(string $attributeName, $value, array $data)
+    {
+        $reflectionMethod = new \ReflectionMethod($this, '__construct');
+
+        foreach ($reflectionMethod->getParameters() as $param) {
+            if ($param->getName() !== $attributeName) {
+                continue;
+            }
+
+            $legacyValue = $data[$attributeName] ?? null;
+
+            if (null === $legacyValue || $param->getDefaultValue() !== $value) {
```

Comment:
We should check if the argument was provided, regardless its value is the same as the default one.

Reply:
The behavior was updated in order to address this comment.